src/matrix.ts

```diff
@@ -60,74 +59,75 @@ export class SparseMatrix {
   private checkDims(row: number, col: number) {
     const withinBounds = row < this.nRows && col < this.nCols;
     if (!withinBounds) {
-      throw new Error('array index out of bounds');
+      throw new Error('row and/or col specified outside of matrix dimensions');
     }
   }
 
   set(row: number, col: number, value: number) {
     this.checkDims(row, col);
     const key = this.makeKey(row, col);
     if (!this.entries.has(key)) {
-      this.rows.push(row);
-      this.cols.push(col);
-      this.values.push(value);
-      this.entries.set(key, this.values.length - 1);
+      this.entries.set(key, { value, row, col });
     } else {
-      const index = this.entries.get(key)!;
-      this.values[index] = value;
+      this.entries.get(key)!.value = value;
     }
   }
 
   get(row: number, col: number, defaultValue = 0) {
     this.checkDims(row, col);
     const key = this.makeKey(row, col);
     if (this.entries.has(key)) {
-      const index = this.entries.get(key)!;
-      return this.values[index];
+      return this.entries.get(key)!.value;
     } else {
       return defaultValue;
     }
   }
 
+  getAll(): { value: number; row: number; col: number }[] {
```

Comment:
OK, so I've figured out that the "integration" tests are failing because the tests expect construction of the `head` and `tail` arrays in `initializeSimplicialSetEmbedding` to be ordered by row then by column. This code is technically working (all of the clustering tests pass), but I'd rather keep the test data as is, so here's a reasonable compromise: just sort the entries returned by `getAll` (which hopefully shouldn't add too much overhead)

```typescript
  getAll(ordered = true): { value: number; row: number; col: number }[] {
    const rowColValues: Entry[] = [];
    this.entries.forEach(value => {
      rowColValues.push(value);
    });
    if (ordered) {
      rowColValues.sort((a, b) => {
        if (a.row === b.row) {
          return a.col - b.col;
        } else {
          return a.row - b.row;
        }
      });
    }

    return rowColValues;
  }
```